pallets/domains/src/lib.rs

```diff
@@ -0,0 +1,583 @@
+        #[pallet::weight(<T as Config>::WeightInfo::reserve_domains(T::DomainsInsertLimit::get()))]
+        pub fn reserve_words(
+            origin: OriginFor<T>,
+            words: Vec<DomainName<T>>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+
+            let words_len = words.len();
+            Self::ensure_insert_limit_not_reached(words_len)?;
+
+            let inserted_words_count = Self::insert_domains(
+                words,
+                Self::ensure_domain_contains_valid_chars,
+                |domain| ReservedWords::<T>::insert(domain, true),
+            )?;
+
+            Self::deposit_event(Event::DomainsReserved(inserted_words_count));
+            Ok((
+                Some(<T as Config>::WeightInfo::reserve_domains(inserted_words_count)),
+                Pays::No,
+            ).into())
+        }
+
+        /// Add a set of top level domains to allowed list.
+        #[pallet::weight(<T as Config>::WeightInfo::reserve_domains(T::DomainsInsertLimit::get()))]
+        pub fn add_tld(
+            origin: OriginFor<T>,
+            domains: Vec<DomainName<T>>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+
+            let domains_len = domains.len();
+            Self::ensure_insert_limit_not_reached(domains_len)?;
+
+            let inserted_domains_count = Self::insert_domains(
+                domains,
+                // TODO: maybe add specific check for TLDs
+                |_| Ok(()),
+                |domain| SupportedTlds::<T>::insert(domain, true),
+            )?;
+
+            Self::deposit_event(Event::NewTldsSupported(inserted_domains_count));
+            Ok((
+                Some(<T as Config>::WeightInfo::reserve_domains(inserted_domains_count)),
+                Pays::No,
+            ).into())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn do_register_domain(
+            owner: T::AccountId,
+            full_domain: DomainName<T>,
+            content: Content,
+            expires_in: <T as frame_system::pallet::Config>::BlockNumber,
+            reserve_deposit: ReserveDeposit,
+        ) -> DispatchResult {
+            ensure!(!expires_in.is_zero(), Error::<T>::ZeroReservationPeriod);
+            ensure!(
+                expires_in <= T::RegistrationPeriodLimit::get(),
+                Error::<T>::TooBigRegistrationPeriod,
+            );
+
+            // Note that while upper and lower case letters are allowed in domain
+            // names, domain names are not case-sensitive. That is, two names with
+            // the same spelling but different cases will be treated as identical.
+            let domain_lc = Self::lower_domain_then_bound(&full_domain);
+
+            Self::ensure_valid_domain(&domain_lc)?;
+
+            let domain_part = domain_lc.split(IS_DOT_CHAR).next().unwrap();
+            let domain_part_bounded = Self::lower_domain_then_bound(domain_part);
```

Comment:
I suggestet we have a function to get a vector of domain lables, and then we pass the vecor to (`ensure_valid_domain`). then we check herer for the reserved word.

in the current code we do split twice.

Reply:
Good point to think on, when refactoring.